src/spaces/openstreetmap.jl

```diff
@@ -0,0 +1,402 @@
+    )
+end
+
+const OSMSpace = OpenStreetMapSpace
+
+struct OSMPos
+    p::Tuple{Int, Int, Float64}
+end
+Base.getindex(o::OSMPos, i::Int) = o.p[i]
+
+"""
+    OSMPos(start::Int, finish::Int = start, road::Float64 = 0)
+
+A helper to provide the correct `pos` format for [`AbstractAgent`](@ref) structs which
+are used in conjuction with [`OpenStreetMapSpace`](@ref).
+It represents the position of the agent on a road connecting the two nodes of the map
+`start, finish` having already covered `road` meters along the road.
+"""
+OSMPos(start::Int, finish::Int = src, p::Float64 = 0.0) = OSMPos((start, finish, p))
+
+#######################################################################################
+# Custom functions for OSMSpace
+#######################################################################################
+
+# EXPORTED
+"""
+    osm_random_direction(model::ABM{A,OpenStreetMapSpace})
+
+Similar to `random_position`, but rather than providing only intersections, this method
+returns a location somewhere on a road heading in a random direction.
+"""
+function osm_random_direction(model::ABM{<:AbstractAgent,<:OpenStreetMapSpace})
+    edge = rand(model.space.edges)
+    OSMPos(edge.I..., rand() * model.space.m.w[edge])
+end
+
+"""
+    osm_plan_route(start, finish, model::ABM{A,OpenStreetMapSpace})
+
+Generate a list of intersections between the `start` and `finish` positions on the map.
+"""
+function osm_plan_route(start::Int, finish::Int, model::ABM{A,OpenStreetMapSpace}) where {A}
+    #TODO: Expand to allow 'fastest_route' as well
+    route =
+        shortest_route(model.space.m, model.space.m.n[start], model.space.m.n[finish])[1]
+    map(p -> getindex(model.space.m.v, p), route)
+end
+
+"""
+    osm_map_coordinates(agent, model::ABM{A,OpenStreetMapSpace})
+
+Return a set of coordinates for an agent on the underlying map. Useful for plotting.
+"""
+function osm_map_coordinates(agent, model)
+    if agent.pos[1] != agent.pos[2]
+        start = get_ENU(agent.pos[1], model)
+        finish = get_ENU(agent.pos[2], model)
+        travelled = agent.pos[3] / Agents.osm_road_length(agent.pos, model)
+        (
+            getX(start) * (1 - travelled) + getX(finish) * travelled,
+            getY(start) * (1 - travelled) + getY(finish) * travelled,
+        )
+    else
+        position = get_ENU(agent.pos[1], model)
+        (getX(position), getY(position))
+    end
+end
+
+"""
+    osm_road_length(pos::OSMPos, model)
+    osm_road_length(start::Int, finish::Int, model)
+
+Returns the distance travelled between two intersections.
+"""
+osm_road_length(pos::OSMPos, model::ABM{<:AbstractAgent,<:OpenStreetMapSpace}) =
+    osm_road_length(pos[1], pos[2], model)
+osm_road_length(p1::Int, p2::Int, model::ABM{<:AbstractAgent,<:OpenStreetMapSpace}) =
+    model.space.m.w[p1, p2]
+
+#HELPERS, NOT EXPORTED
+
+get_ENU(pos::Int, model) = model.space.m.nodes[model.space.m.n[pos]]
```

Comment:
No idea what this means. ENU? Can we rename it to something clearer and more verbose? Since it is internal its length doesn't matter much.

Reply:
It's an OSM thing. I don't know. I just needed it to convert the underlying structure. Will dig deeper into what it is....